core/src/toga/window.py

```diff
@@ -256,19 +292,34 @@ def visible(self, visible: bool) -> None:
 
     @property
     def on_close(self) -> OnCloseHandler:
-        """The handler to invoke before the window is closed."""
+        """The handler to invoke before the window is closed in response to a user
+        action.
+
+        If the handler returns ``False``, the request to close the window will be
+        cancelled.
+        """
         return self._on_close
 
     @on_close.setter
     def on_close(self, handler: OnCloseHandler | None) -> None:
         def cleanup(window: Window, should_close: bool) -> None:
-            if should_close:
+            if should_close or handler is None:
                 window.close()
 
         self._on_close = wrapped_handler(self, handler, cleanup=cleanup)
 
     def close(self) -> None:
+        """Close the window.
+
+        This *does not* invoke the ``on_close`` handler; the window will be immediately
+        and unconditionally closed.
+
+        Once a window has been closed, it *cannot* be reused. The behavior of any method
+        on a :class:`~toga.Window` instance after it has been closed is undefined.
+        """
         self.app.windows -= self
+        for widget in self.widgets:
+            self.app.widgets.remove(widget.id)
```

Comment:
Now that the widget registry is weak, the intention is to keep widgets associated with the app for their entire life. In this case, we no longer need to explicitly remove widgets from the registry on window close, right? I'll remove this part again.